server/src/main/java/com/linecorp/centraldogma/server/internal/storage/DirectoryBasedStorageManager.java

```diff
@@ -347,11 +419,51 @@ private static boolean isValidChildName(String name) {
             return false;
         }
 
-        return !name.endsWith(SUFFIX_REMOVED);
+        return !name.endsWith(SUFFIX_REMOVED) && !name.endsWith(SUFFIX_PURGED);
     }
 
     @Override
     public String toString() {
         return Util.simpleTypeName(getClass()) + '(' + rootDir + ')';
     }
+
+    private final class StorageRemovalManager {
+
+        private static final String REMOVAL_TIMESTAMP_NAME = ".removal.timestamp";
+
+        void mark(File file) {
+            final File removal = new File(file, REMOVAL_TIMESTAMP_NAME);
+            final String timestamp = DateTimeFormatter.ISO_INSTANT.format(Instant.now());
+            try {
+                Files.write(removal.toPath(), timestamp.getBytes(StandardCharsets.UTF_8));
+            } catch (IOException e) {
+                throw new StorageException(
+                        "failed to write a removal timestamp " + childTypeName + ": " + removal);
+            }
+        }
+
+        void unmark(File file) {
+            final File removal = new File(file, REMOVAL_TIMESTAMP_NAME);
+            if (removal.exists()) {
+                if (!removal.delete()) {
+                    logger.warn("Failed to delete a removal timestamp {}: {}", childTypeName, removal);
+                }
+            }
+        }
+
+        Instant readRemoval(File file) {
+            final File removal = new File(file, REMOVAL_TIMESTAMP_NAME);
+            if (!removal.exists()) {
+                return Instant.ofEpochMilli(file.lastModified());
+            }
+            try {
+                final String timestamp = new String(Files.readAllBytes(removal.toPath()),
+                                                    StandardCharsets.UTF_8);
+                return Instant.from(DateTimeFormatter.ISO_INSTANT.parse(timestamp));
+            } catch (IOException e) {
```

Comment:
`Exception`? (We should catch a `DateTimeParseException`.)

Reply:
Thanks, I will keep in mind.